### PR TITLE
integrate: add Trapezoidal implementation

### PR DESCRIPTION
**Pull request by sbinet**

### CODE REVIEW

**Comment by vladimir-ch** [comment]
I was busy last week so I'm sorry for not asking before this was moved to here from stat: does AUC really belong to quad? The package description says that "...quad provides numerical evaluation of definite integrals of single-variable _functions_...". AUC here takes a set of samples (no curve involved), interpolates them piece-wise linearly (arbitrary), and then computes _exactly_ the definite integral of this interpolation. Alternative interpretation, it takes a set of samples of some unknown function, does no interpolation, and computes _approximately_ the definite integral of the sampled function. This might be ok for applications and indeed AUC seems to be a very application-oriented term whose definition (or interpretation) varies with each application. I would like to keep quad oriented on what the description says (quadrature in the numerical mathematics sense). I feel a bit bad for saying that now but I would argue that it belongs more to stat.

**Comment by btracey** [comment]
This is quadrature though, or at least something like it. It's trying to find the integral of a function. The difference being the samples are generated ahead of time. I agree there's not that clear of a line (Monte Carlo is also finding an integral), but I don't see anything statistical about the algorithm. 

It seems to me like there are two basic interfaces for quad. One is that you have a function and you want its integral, the other is that you have a set of samples, and you want to estimate the integral knowing those. Those feel like clean concepts to me. I'm not as sure where they both should live and how to avoid name clashes. In any event, I don't think it belongs in the base stat package. As you say, the linear interpolation is somewhat arbitrary, and there are a whole bunch of ways one could do it (hence an interface).

**Comment by sbinet** [comment]
SGTM: I could indeed introduce some interface:

```go
type SampleBasedIntegral interface {
   Integral(x, y []float64) float64
}
```

(names to be refined)

and then introduce:

```go
type TrapezoidalIntegral struct {}
func (TrapezoidalIntegral) Integral(x,y []float64) float64 { ... }
```

so `AUC` would take a value of that interface (`nil`-able and defaults to `TrapezoidalIntegral`).

what do you think?

**Comment by vladimir-ch** [comment]
I agree, there's nothing statistical about the algorithm but googling shows that the term AUC is used almost exclusively in statistical context. So at least the function should have a different name and the link from the doc comment should be removed.

**Comment by btracey** [comment]
Yes, I agree the name should change, but how? This basically is the trapezoidal rule, but what's the best way to distinguish this from 
````
Trapezoidal(f []float64, points int)
````
(or whatever). There are three kinds of routines. Non-adaptive (where the number of samples is set), Adaptive (where the user sets an accuracy), and Set (where the samples is set). I was thinking about this a while ago and I think Non-adaptive and Adaptive can be wrapped into a single "Method", but obviously Set is very different. Should they be two different packages? Or is there a clean way to do the names to separate the concepts.

**Comment by sbinet** [comment]
`scipy` has a special module for them: http://docs.scipy.org/doc/scipy-0.17.0/reference/integrate.html

we could have:

```go
package integrate

func Trapz(x, y []float64) float64 {...}
func Simpson(x, y []float64) float64 {...}
func Romberg(x, y []float64) float64 {...}
```

(it's relatively easy to generate `x []float64` from `floats.Span`. It could be made easier, though.)

`integrate` could be a sibling of `quad` or child.

**Comment by kortschak** [comment]
Please not "Trapz".

**Comment by btracey** [comment]
Here is a suggestion plus reasons (that are incorporated from the discussion)

- It seems like there should be two packages. Computing an integral given direct access to the function vs. computing an integral only given a specific list of evaluations are nicely separable concepts, and pretty different in character. In particular, adaptive algorithms aren't possible given a fixed set of samples. 
- In addition to the above, there are likely a lot of name overlaps, and so having everything in one place could be ugly.
- Proposal: Change the base package from `quad` to `integrate`. The current quad package lives at github.com/gonum/integrate/quad. As @sbinet proposes, integration given function evaluations lives in the 'base' integrate package.
- Reasoning: At some level, the code in integrate is the base concept. There are really three stages: 1) Generate sample locations 2) Evaluate the function 3) Estimate the integral. The new integrate package is this third step
- Quad links all three of these steps together. At the present, this is only for convenience (i.e. parallelism, etc.), and in the long run adaptive quadrature is enough of a benefit to exist in its present form.
- That said, this split allows for new subpackages, for example, integrate/fixedquad (or whatever) which contains the routines for generating these locations. Thus, the current quad fixed code could be something like
````
xs, ws := fixedquad.Legendre(50)
var integral float64
for i, x := range xs{
     integral += f(x) * ws[i]
}
return integral
````
Though the adaptive algorithms would be more complicated. 
- There are also more elaborate algorithms (sparse grids, etc.) where the user may want to generate the points separately from evaluating the function (for example as a list of locations to send out to a cluster). This new organization provides a clear way to do have this separation.

**Comment by sbinet** [comment]
SGTM.
(`s/Trapz/Trapezoidal/g` included)

**Comment by vladimir-ch** [comment]
I was thinking about quad a while back, the following is an unsorted set of thoughts that have not escaped my mind yet.

I like the base package to be `integrate`.

I'm not sure about introducing many additional subpackages (see matrix), although `quad` could host the quadrature rules and thus make it more convenient to type, while `integrate` could have the integration algorithms.

In my proposal for `ode`, @btracey suggested to define the top level functions on the type of problem they solve. Wouldn't the same idea work here better than splitting based on fixed/adaptive integration? Perhaps split on finite/infinite bounds: `integrate.Finite()`, `integrate.Infinite()`. Or on smooth function/discontinuous function/samples: `integrate.Smooth()`, `integrate.Discontinuous()`, `integrate.Samples()`. This delegates the fixedness/adaptivity of the quadrature to the quadrature rules which all would have to implement a common interface.

There are different kinds of adaptivity. For example, rules that can generate a sequence of subdivisions where all points from the previous level are reused (with different weights) and new points are added (=nested/embedded quadrature rules, e.g. Gauss-Kronrod). Another example are algorithms that bisect the integration interval until the error estimate is small.

Just like we don't have `optimize.LBFGS(problem, xinit, settings)`, I think we should not have `integrate.Trapezoidal(x,y)` et al.
I would prefer to see something like
`integrate.Samples(x, y, &quad.Trapezoidal{})`
A (small) problem is that `Trapezoidal` can be used also to integrate smooth functions in which case it needs the number of subintervals (not that I'm a big fan of such rules, you don't know how to choose N and have no estimate how accurate the result is):
`integrate.Function(f, a, b, &quad.Trapezoidal{N: 100})`
When used in `Samples`, `N` would be ignored which may be not be considered a nice design.

**Comment by btracey** [comment]
I am fully on board with high-level functions with a "method" argument. I'm also reasonably sure that adaptive and non-adaptive can be under the same interface. I tried this at one point, but it's a bit tricky with parallelism. I think I like "integrate.Function" and "integrate.Samples".

The name clashes I mentioned are more in the method names than the high-level names. I'm not yet sure how we mark "Integrate this function with the Trapezoidal rule" vs. "Integrate these samples using a Trapezoidal rule".

**Comment by vladimir-ch** [comment]
> I think I like ... "integrate.Samples".

Also `integrate.Discrete` or `integrate.DiscreteData`?

**Comment by sbinet** [comment]
ping?

File: trapezoidal.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import "sort"
+
+// Trapezoidal returns the integral value of the function sampled by the set of points (x,y).
+// The integral value is approximated by a simple trapezoidal rule.
+//
+// The x and y input data points must be sorted and of equal length (>1).
+//
+// More details about the trapezoidal rule are available at:
+// https://en.wikipedia.org/wiki/Trapezoidal_rule
+func Trapezoidal(x, y []float64) float64 {
```
Comment (btracey) [inline trapezoidal.go]: change to x and f?

File: trapezoidal.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import "sort"
+
+// Trapezoidal returns the integral value of the function sampled by the set of points (x,y).
```
Comment (btracey) [inline trapezoidal.go]: My wording:

`Trapezoidal estimates the integral of a function from a set of evaluations of the function using the trapezoidal rule. It estimates
  \int_a^b f(x) dx
where f[i] is the value of the function at x[i]. The trapezoidal rule makes piecewise trapezoidal approximations to the function, and estimates 
  \int_x[i]^x[i+1] f(x)dx
as 
  (x[i+1] - x[i]) * (f[i] + f[i+1])/2
More details on the trapezoidal rule can be found at https://en.wikipedia.org/wiki/Trapezoidal_rule

The x and y input data points must be sorted and of equal length (>1).`

File: trapezoidal_test.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+package integrate
+
+import "testing"
+
+func TestTrapezoidal(t *testing.T) {
+	for i, test := range []struct {
+		x, y []float64
+		want float64
+	}{
+		{
```
Comment (btracey) [inline trapezoidal_test.go]: Could you also add a nonlinear function test?

File: trapezoidal.go (hunk 2 above)
Comment (sbinet) [inline trapezoidal.go]: done

File: trapezoidal.go (hunk 1 above)
Comment (sbinet) [inline trapezoidal.go]: done

File: trapezoidal_test.go (hunk 3 above)
Comment (sbinet) [inline trapezoidal_test.go]: done

File: trapezoidal.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import "sort"
+
+// Trapezoidal estimates the integral of a function from a set of evaluations of
```
Comment (vladimir-ch) [inline trapezoidal.go]: I would modify it slightly to
```
// Trapezoidal estimates the integral of a function f
//  \int_a^b f(x) dx
// from a set of evaluations of
// the function using the trapezoidal rule.
// The trapezoidal rule makes
// piecewise linear approximations to the function, and estimates
//  \int_x[i]^x[i+1] f(x) dx
// as
//  (x[i+1] - x[i]) * (f[i] + f[i+1])/2
// where f[i] is the value of the function at x[i].
```

Comment (sbinet) [inline trapezoidal.go]: SGTM. (done)

File: trapezoidal.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package integrate
+
+import "sort"
+
+// Trapezoidal estimates the integral of a function f
```
Comment (btracey) [inline trapezoidal.go]: Could you clean up the line breaks in the comment so it reads more smoothly?

File: trapezoidal.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+// piecewise linear approximations to the function, and estimates
+//  \int_x[i]^x[i+1] f(x) dx
+// as
+//  (x[i+1] - x[i]) * (f[i] + f[i+1])/2
+// where f[i] is the value of the function at x[i].
+// More details on the trapezoidal rule can be found at:
+// https://en.wikipedia.org/wiki/Trapezoidal_rule
+//
+// The x input data points must be sorted.
+// The x and f slices must be of equal length (>1).
```
Comment (btracey) [inline trapezoidal.go]: must be of equal length and have length >= 1.

File: trapezoidal.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+// The trapezoidal rule makes
+// piecewise linear approximations to the function, and estimates
+//  \int_x[i]^x[i+1] f(x) dx
+// as
+//  (x[i+1] - x[i]) * (f[i] + f[i+1])/2
+// where f[i] is the value of the function at x[i].
+// More details on the trapezoidal rule can be found at:
+// https://en.wikipedia.org/wiki/Trapezoidal_rule
+//
+// The x input data points must be sorted.
```
Comment (btracey) [inline trapezoidal.go]: Add comment that stat.SortWeighted can be used to sort `f` and `x` together?

**Comment by btracey** [comment]
LGTM with comment fixes. Sorry, didn't see you had uploaded more changes.

File: trapezoidal.go (hunk 5 above)
Comment (sbinet) [inline trapezoidal.go]: done.

File: trapezoidal.go (hunk 6 above)
Comment (sbinet) [inline trapezoidal.go]: done.

File: trapezoidal.go (hunk 7 above)
Comment (sbinet) [inline trapezoidal.go]: done.

**Comment by sbinet** [comment]
PTAL.

**Comment by sbinet** [comment]
(if everybody is happy with it, I'll squash and push)

**Comment by sbinet** [comment]
squashed and pushed.

**Pull request merged**